MDEV-35353 set correct number of rows_examined for union all queries

bsrikanth-mariadb opened the pull request:
Pull request created in: https://jira.mariadb.org/browse/MDEV-35353

spetrunia commented:
I think the commit comment
```
For all union_all, as well as union queries, number of rows_examined
was tracked and updated correctly. However, they are only being
written to the slow query log in the case of UNION queries, but not
for UNION_ALL queries.
```
is very misleading.

What happens is: in `st_select_lex_unit::exec()`  there is code to "take away" the value from thd->m_examined_row_count.  For each part of the UNION, we do:

```cpp
            examined_rows+= thd->get_examined_row_count();
            thd->set_examined_row_count(0);
```

then, at the end we add back the number we've removed:
```cpp
          thd->inc_examined_row_count(examined_rows);
```
I am not fully sure what is the goal of this.

Anyhow, the statement that 
> number of rows_examined was tracked and updated correctly.

is incorrect: before the patch, a query  `select * from three_rows union select * from four_rows`  would NOT compute rows_examined= 3+4=7.

The statement that
>  However, they are only being written to the slow query log in the case of UNION queries, but not
for UNION_ALL queries.
also looks odd.  AFAIU the part of the code that writes to slow query log doesn't care whether the query had a UNION or UNION ALL...

spetrunia commented:
@bsrikanth-mariadb :
 * Any idea why the  "take away" and "add back" logic is there?  (if we remove both, is there any difference?)
 * Please adjust the commit comment to be more clear.
 * (More cosmetic input to come)

spetrunia commented on mysql-test/main/log_state.result:
```diff
@@ -313,6 +313,86 @@ disconnect con2;
+SET GLOBAL long_query_time = 0.0001;
+TRUNCATE TABLE mysql.slow_log;
+create table t1(a int, b int);
+insert into t1 select seq, seq from seq_1_to_10000;
```
Do we need 10K rows here? 
I would use table t1 with say 20 rows and table t2 with say 5 rows, and then make sure that rows_examined=20+5=25 (or 26 in some cases)

bsrikanth-mariadb replied:
Not actually. Thought if number of rows are less, then does slow query log even get populated. May be we might have to adjust long_query_time.

spetrunia replied:
Yes, we definitely need `long_query_time=0.0`.  Any other value has a risk that some build host will be too fast.

spetrunia commented on sql/sql_union.cc:
```diff
@@ -2435,10 +2435,8 @@ bool st_select_lex_unit::exec()
 
       fake_select_lex->table_list.empty();
       if (likely(!saved_error))
```
So we end up with
```cpp
      if (likely(!saved_error))
        thd->limit_found_rows=
            (ulonglong) table->file->stats.records + add_rows;
```
coding style somewhere says one needs to use { } if the body of if/while/etc statement takes multiple lines, so use
```cpp
      if (likely(!saved_error))
      {
        thd->limit_found_rows=
            (ulonglong) table->file->stats.records + add_rows;
      }
```

bsrikanth-mariadb commented:
> I think the commit comment
> 
> ```
> For all union_all, as well as union queries, number of rows_examined
> was tracked and updated correctly. However, they are only being
> written to the slow query log in the case of UNION queries, but not
> for UNION_ALL queries.
> ```
> 
> is very misleading.
> 
> What happens is: in `st_select_lex_unit::exec()` there is code to "take away" the value from thd->m_examined_row_count. For each part of the UNION, we do:
> 
> ```c++
>             examined_rows+= thd->get_examined_row_count();
>             thd->set_examined_row_count(0);
> ```
> 
> then, at the end we add back the number we've removed:
> 
> ```c++
>           thd->inc_examined_row_count(examined_rows);
> ```
> 
> I am not fully sure what is the goal of this.
> 
> Anyhow, the statement that
> 
> > number of rows_examined was tracked and updated correctly.
> 
> is incorrect: before the patch, a query `select * from three_rows union select * from four_rows` would NOT compute rows_examined= 3+4=7.
> 
> The statement that
> 
> > However, they are only being written to the slow query log in the case of UNION queries, but not
> > for UNION_ALL queries.
> 
> also looks odd. AFAIU the part of the code that writes to slow query log doesn't care whether the query had a UNION or UNION ALL...

changed the commit message

spetrunia reviewed (approved):
Please use `long_query_time=0.0`.  
Ok to push after that.
